# Pandas Grouping and Aggregating

In [1]:
import pandas as pd

In [6]:
df = pd.read_csv("data/survey_results_public.csv")

pd.set_option("display.max_columns", 85)
pd.set_option("display.max_rows", 85)

In [7]:
df1 = pd.read_csv("csv_file.csv")

In [8]:
students = {
    "first_name" : ["Shashi", "Meet", "Anshuman", "Prashant", "Vikhil"],
    "last_name" : ["Prajapati", "Thakar", "Tiwari", "Vasala", "Sallagargi"],
    "Roll-no" : [46, 64, 65, 67, 50], 
    "email" : ["shashiprajapati38@gmail.com", "meet@gmail.com", "anshuman@gmail.com", "prashant@gmail.com", "vikhil@gmail.com"]
}

In [9]:
df2 = pd.DataFrame(students)

In [11]:
df["ConvertedComp"].head(10)

0         NaN
1         NaN
2      8820.0
3     61000.0
4         NaN
5    366420.0
6         NaN
7         NaN
8     95179.0
9     13293.0
Name: ConvertedComp, dtype: float64

In [12]:
df["ConvertedComp"].median()      # Median Salary

57287.0

In [13]:
df.median()    # it searches for all the columns containing nos, and finds the median of those

Respondent       44442.0
CompTotal        62000.0
ConvertedComp    57287.0
WorkWeekHrs         40.0
CodeRevHrs           4.0
Age                 29.0
dtype: float64

In [15]:
df.describe()                          # for complete dataset

Respondent     CompTotal  ConvertedComp   WorkWeekHrs    CodeRevHrs  \
count  88883.000000  5.594500e+04   5.582300e+04  64503.000000  49790.000000   
mean   44442.000000  5.519014e+11   1.271107e+05     42.127197      5.084308   
std    25658.456325  7.331926e+13   2.841523e+05     37.287610      5.513931   
min        1.000000  0.000000e+00   0.000000e+00      1.000000      0.000000   
25%    22221.500000  2.000000e+04   2.577750e+04     40.000000      2.000000   
50%    44442.000000  6.200000e+04   5.728700e+04     40.000000      4.000000   
75%    66662.500000  1.200000e+05   1.000000e+05     44.750000      6.000000   
max    88883.000000  1.000000e+16   2.000000e+06   4850.000000     99.000000   

                Age  
count  79210.000000  
mean      30.336699  
std        9.178390  
min        1.000000  
25%       24.000000  
50%       29.000000  
75%       35.000000  
max       99.000000

In [16]:
df["ConvertedComp"].count()

55823

In [17]:
df["Hobbyist"]

0        Yes
1         No
2        Yes
3         No
4        Yes
        ... 
88878    Yes
88879     No
88880     No
88881     No
88882    Yes
Name: Hobbyist, Length: 88883, dtype: object

In [18]:
df["Hobbyist"].value_counts()

Yes    71257
No     17626
Name: Hobbyist, dtype: int64

In [19]:
df["SocialMedia"]

0          Twitter
1        Instagram
2           Reddit
3           Reddit
4         Facebook
           ...    
88878      YouTube
88879          NaN
88880          NaN
88881          NaN
88882     WhatsApp
Name: SocialMedia, Length: 88883, dtype: object

In [20]:
df["SocialMedia"].value_counts()

Reddit                      14374
YouTube                     13830
WhatsApp                    13347
Facebook                    13178
Twitter                     11398
Instagram                    6261
I don't use social media     5554
LinkedIn                     4501
WeChat 微信                     667
Snapchat                      628
VK ВКонта́кте                 603
Weibo 新浪微博                     56
Youku Tudou 优酷                 21
Hello                          19
Name: SocialMedia, dtype: int64

In [21]:
df["SocialMedia"].value_counts(normalize=True) # in percentage (out of 1)

Reddit                      0.170233
YouTube                     0.163791
WhatsApp                    0.158071
Facebook                    0.156069
Twitter                     0.134988
Instagram                   0.074150
I don't use social media    0.065777
LinkedIn                    0.053306
WeChat 微信                   0.007899
Snapchat                    0.007437
VK ВКонта́кте               0.007141
Weibo 新浪微博                  0.000663
Youku Tudou 优酷              0.000249
Hello                       0.000225
Name: SocialMedia, dtype: float64

Get with specific countries

In [22]:
df["Country"]

0                United Kingdom
1        Bosnia and Herzegovina
2                      Thailand
3                 United States
4                       Ukraine
                  ...          
88878                    Canada
88879                       NaN
88880                       NaN
88881                       NaN
88882                     Spain
Name: Country, Length: 88883, dtype: object

In [23]:
df["Country"].value_counts()

United States            20949
India                     9061
Germany                   5866
United Kingdom            5737
Canada                    3395
                         ...  
Brunei Darussalam            1
Papua New Guinea             1
Dominica                     1
Sao Tome and Principe        1
Saint Kitts and Nevis        1
Name: Country, Length: 179, dtype: int64

In [25]:
df.groupby(["Country"])

In [26]:
country_grp = df.groupby(["Country"])

In [28]:
country_grp.get_group("India")

Respondent                              MainBranch Hobbyist  \
7               8             I code primarily as a hobby      Yes   
9              10          I am a developer by profession      Yes   
14             15  I am a student who is learning to code      Yes   
49             50          I am a developer by profession      Yes   
64             65          I am a developer by profession      Yes   
...           ...                                     ...      ...   
88808       77339                                     NaN      Yes   
88825       79795                                     NaN      Yes   
88852       83862                                     NaN      Yes   
88853       84299                                     NaN      Yes   
88864       86012                                     NaN      Yes   

                                             OpenSourcer  \
7                                Less than once per year   
9                             Once a month or more often   
14                                                 Never   
49                            Once a month or more often   
64                                                 Never   
...                                                  ...   
88808                            Less than once per year   
88825  Less than once a month but more than once per ...   
88852                         Once a month or more often   
88853                                              Never   
88864                         Once a month or more often   

                                              OpenSource  \
7      OSS is, on average, of HIGHER quality than pro...   
9      OSS is, on average, of HIGHER quality than pro...   
14     OSS is, on average, of HIGHER quality than pro...   
49     OSS is, on average, of LOWER quality than prop...   
64                                                   NaN   
...                                                  ...   
88808  OSS is, on average, of HIGHER quality than pro...   
88825  OSS is, on average, of HIGHER quality than pro...   
88852  OSS is, on average, of HIGHER quality than pro...   
88853  The quality of OSS and closed source software ...   
88864  OSS is, on average, of HIGHER quality than pro...   

                               Employment Country         Student  \
7      Not employed, but looking for work   India             NaN   
9                      Employed full-time   India              No   
14     Not employed, but looking for work   India  Yes, full-time   
49                     Employed full-time   India              No   
64                     Employed full-time   India              No   
...                                   ...     ...             ...   
88808                  Employed full-time   India  Yes, full-time   
88825                  Employed full-time   India              No   
88852                  Employed full-time   India  Yes, full-time   
88853                  Employed full-time   India  Yes, full-time   
88864                  Employed full-time   India              No   

                                                 EdLevel  \
7               Bachelor’s degree (BA, BS, B.Eng., etc.)   
9            Master’s degree (MA, MS, M.Eng., MBA, etc.)   
14     Secondary school (e.g. American high school, G...   
49              Bachelor’s degree (BA, BS, B.Eng., etc.)   
64              Bachelor’s degree (BA, BS, B.Eng., etc.)   
...                                                  ...   
88808           Bachelor’s degree (BA, BS, B.Eng., etc.)   
88825           Bachelor’s degree (BA, BS, B.Eng., etc.)   
88852           Bachelor’s degree (BA, BS, B.Eng., etc.)   
88853           Bachelor’s degree (BA, BS, B.Eng., etc.)   
88864        Master’s degree (MA, MS, M.Eng., MBA, etc.)   

                                          UndergradMajor  \
7      Computer science, computer engineering, or sof...   
9                                                    N

In [29]:
filt1 = df["Country"] == "India"

In [32]:
df.loc[filt1]

Respondent                              MainBranch Hobbyist  \
7               8             I code primarily as a hobby      Yes   
9              10          I am a developer by profession      Yes   
14             15  I am a student who is learning to code      Yes   
49             50          I am a developer by profession      Yes   
64             65          I am a developer by profession      Yes   
...           ...                                     ...      ...   
88808       77339                                     NaN      Yes   
88825       79795                                     NaN      Yes   
88852       83862                                     NaN      Yes   
88853       84299                                     NaN      Yes   
88864       86012                                     NaN      Yes   

                                             OpenSourcer  \
7                                Less than once per year   
9                             Once a month or more often   
14                                                 Never   
49                            Once a month or more often   
64                                                 Never   
...                                                  ...   
88808                            Less than once per year   
88825  Less than once a month but more than once per ...   
88852                         Once a month or more often   
88853                                              Never   
88864                         Once a month or more often   

                                              OpenSource  \
7      OSS is, on average, of HIGHER quality than pro...   
9      OSS is, on average, of HIGHER quality than pro...   
14     OSS is, on average, of HIGHER quality than pro...   
49     OSS is, on average, of LOWER quality than prop...   
64                                                   NaN   
...                                                  ...   
88808  OSS is, on average, of HIGHER quality than pro...   
88825  OSS is, on average, of HIGHER quality than pro...   
88852  OSS is, on average, of HIGHER quality than pro...   
88853  The quality of OSS and closed source software ...   
88864  OSS is, on average, of HIGHER quality than pro...   

                               Employment Country         Student  \
7      Not employed, but looking for work   India             NaN   
9                      Employed full-time   India              No   
14     Not employed, but looking for work   India  Yes, full-time   
49                     Employed full-time   India              No   
64                     Employed full-time   India              No   
...                                   ...     ...             ...   
88808                  Employed full-time   India  Yes, full-time   
88825                  Employed full-time   India              No   
88852                  Employed full-time   India  Yes, full-time   
88853                  Employed full-time   India  Yes, full-time   
88864                  Employed full-time   India              No   

                                                 EdLevel  \
7               Bachelor’s degree (BA, BS, B.Eng., etc.)   
9            Master’s degree (MA, MS, M.Eng., MBA, etc.)   
14     Secondary school (e.g. American high school, G...   
49              Bachelor’s degree (BA, BS, B.Eng., etc.)   
64              Bachelor’s degree (BA, BS, B.Eng., etc.)   
...                                                  ...   
88808           Bachelor’s degree (BA, BS, B.Eng., etc.)   
88825           Bachelor’s degree (BA, BS, B.Eng., etc.)   
88852           Bachelor’s degree (BA, BS, B.Eng., etc.)   
88853           Bachelor’s degree (BA, BS, B.Eng., etc.)   
88864        Master’s degree (MA, MS, M.Eng., MBA, etc.)   

                                          UndergradMajor  \
7      Computer science, computer engineering, or sof...   
9                                                    N

In [33]:
df.loc[filt1]["SocialMedia"]

7                         YouTube
9                         YouTube
14                        YouTube
49                        YouTube
64                       WhatsApp
                   ...           
88808                     YouTube
88825                   Instagram
88852                     Twitter
88853                    LinkedIn
88864    I don't use social media
Name: SocialMedia, Length: 9061, dtype: object

In [34]:
df.loc[filt1]["SocialMedia"].value_counts()

WhatsApp                    2990
YouTube                     1820
LinkedIn                     955
Facebook                     841
Instagram                    822
Twitter                      542
Reddit                       473
I don't use social media     250
Snapchat                      23
Hello                          5
WeChat 微信                      5
VK ВКонта́кте                  4
Youku Tudou 优酷                 2
Weibo 新浪微博                     1
Name: SocialMedia, dtype: int64

In [37]:
country_grp["SocialMedia"].value_counts().head(50)

Country              SocialMedia             
Afghanistan          Facebook                     15
                     YouTube                       9
                     I don't use social media      6
                     WhatsApp                      4
                     Instagram                     1
                     LinkedIn                      1
                     Twitter                       1
Albania              WhatsApp                     18
                     Facebook                     16
                     Instagram                    13
                     YouTube                      10
                     Twitter                       8
                     LinkedIn                      7
                     Reddit                        6
                     I don't use social media      4
                     Snapchat                      1
                     WeChat 微信                     1
Algeria              YouTube                      42


In [39]:
country_grp["SocialMedia"].value_counts()

Country      SocialMedia             
Afghanistan  Facebook                    15
             YouTube                      9
             I don't use social media     6
             WhatsApp                     4
             Instagram                    1
                                         ..
Zimbabwe     Facebook                     3
             YouTube                      3
             Instagram                    2
             LinkedIn                     2
             Reddit                       1
Name: SocialMedia, Length: 1220, dtype: int64

In [40]:
country_grp["SocialMedia"].value_counts().loc["India"]

SocialMedia
WhatsApp                    2990
YouTube                     1820
LinkedIn                     955
Facebook                     841
Instagram                    822
Twitter                      542
Reddit                       473
I don't use social media     250
Snapchat                      23
Hello                          5
WeChat 微信                      5
VK ВКонта́кте                  4
Youku Tudou 优酷                 2
Weibo 新浪微博                     1
Name: SocialMedia, dtype: int64

In [41]:
country_grp["SocialMedia"].value_counts(normalize=True).loc["India"]

SocialMedia
WhatsApp                    0.342379
YouTube                     0.208405
LinkedIn                    0.109355
Facebook                    0.096301
Instagram                   0.094126
Twitter                     0.062063
Reddit                      0.054162
I don't use social media    0.028627
Snapchat                    0.002634
Hello                       0.000573
WeChat 微信                   0.000573
VK ВКонта́кте               0.000458
Youku Tudou 优酷              0.000229
Weibo 新浪微博                  0.000115
Name: SocialMedia, dtype: float64

In [42]:
country_grp["ConvertedComp"].describe()

count           mean            std  \
Country                                                                     
Afghanistan                            12.0  101953.333333  285995.241798   
Albania                                50.0   21833.700000   31450.766680   
Algeria                                42.0   34924.047619  152933.628015   
Andorra                                 2.0  160931.000000   15458.768450   
Angola                                  1.0    7764.000000            NaN   
...                                     ...            ...            ...   
Venezuela, Bolivarian Republic of...   43.0   14581.627907   24218.759814   
Viet Nam                               87.0   17233.436782   19921.575978   
Yemen                                   6.0   16909.166667   21847.429143   
Zambia                                  8.0   10075.375000   13200.366434   
Zimbabwe                               15.0   34046.666667   44166.969990   

                                           min       25%       50%        75%  \
Country                                                                         
Afghanistan                                0.0    1476.0    6222.0   17955.00   
Albania                                 1320.0    7314.0   10818.0   23052.00   
Algeria                                    0.0    4701.0    7878.0   14391.00   
Andorra                               150000.0  155465.5  160931.0  166396.50   
Angola                                  7764.0    7764.0    7764.0    7764.00   
...                                        ...       ...       ...        ...   
Venezuela, Bolivarian Republic of...       0.0    1950.0    6384.0   18000.00   
Viet Nam                                 200.0    7236.0   11892.0   18000.00   
Yemen                                    799.0    4437.0   11940.0   14115.00   
Zambia                                   400.0    2103.0    5040.0   12734.25   
Zimbabwe                                 900.0   12000.0   19200.0   36000.00   

                                            max  
Country                                          
Afghanistan                           1000000.0  
Albania                                187668.0  
Algeria                               1000000.0  
Andorra                                171862.0  
Angola                                   7764.0  
...                                         ...  
Venezuela, Bolivarian Republic of...   137484.0  
Viet Nam                               140000.0  
Yemen                                   60000.0  
Zambia                                  40524.0  
Zimbabwe                               180000.0  

[179 rows x 8 columns]

In [43]:
country_grp["ConvertedComp"].median()

Country
Afghanistan                               6222.0
Albania                                  10818.0
Algeria                                   7878.0
Andorra                                 160931.0
Angola                                    7764.0
                                          ...   
Venezuela, Bolivarian Republic of...      6384.0
Viet Nam                                 11892.0
Yemen                                    11940.0
Zambia                                    5040.0
Zimbabwe                                 19200.0
Name: ConvertedComp, Length: 179, dtype: float64

In [44]:
country_grp["ConvertedComp"].median().loc["India"]

10080.0

In [45]:
country_grp["ConvertedComp"].agg(["median", "mean"])

median           mean
Country                                                      
Afghanistan                             6222.0  101953.333333
Albania                                10818.0   21833.700000
Algeria                                 7878.0   34924.047619
Andorra                               160931.0  160931.000000
Angola                                  7764.0    7764.000000
...                                        ...            ...
Venezuela, Bolivarian Republic of...    6384.0   14581.627907
Viet Nam                               11892.0   17233.436782
Yemen                                  11940.0   16909.166667
Zambia                                  5040.0   10075.375000
Zimbabwe                               19200.0   34046.666667

[179 rows x 2 columns]

In [46]:
country_grp["ConvertedComp"].agg(["median", "mean"]).loc["India"]

median    10080.000000
mean      28057.664916
Name: India, dtype: float64

In [53]:
filt2 = df["Country"] == "India"
df.loc[filt2]["LanguageWorkedWith"]

7        Bash/Shell/PowerShell;C;C++;HTML/CSS;Java;Java...
9                            C#;Go;JavaScript;Python;R;SQL
14       Assembly;Bash/Shell/PowerShell;C;C++;HTML/CSS;...
49       Bash/Shell/PowerShell;C;C++;HTML/CSS;Java;Java...
64                         Assembly;C;C++;C#;HTML/CSS;Java
                               ...                        
88808                                            Other(s):
88825                 Bash/Shell/PowerShell;Python;SQL;VBA
88852    Assembly;C;C++;HTML/CSS;Java;JavaScript;Object...
88853     HTML/CSS;Java;JavaScript;Python;Swift;TypeScript
88864        Bash/Shell/PowerShell;C++;HTML/CSS;JavaScript
Name: LanguageWorkedWith, Length: 9061, dtype: object

In [54]:
df.loc[filt2]["LanguageWorkedWith"].str.contains("Python")

7         True
9         True
14       False
49        True
64       False
         ...  
88808    False
88825     True
88852    False
88853     True
88864    False
Name: LanguageWorkedWith, Length: 9061, dtype: object

Now Count how many people know python in India

In [55]:
df.loc[filt2]["LanguageWorkedWith"].str.contains("Python").sum()

3105

In [56]:
country_grp["LanguageWorkedWith"].apply(lambda x: x.str.contains("Python"))

0         True
1         True
2        False
3         True
4         True
         ...  
88878    False
88879      NaN
88880      NaN
88881      NaN
88882    False
Name: LanguageWorkedWith, Length: 88883, dtype: object

In [57]:
country_grp["LanguageWorkedWith"].apply(lambda x: x.str.contains("Python").sum())

Country
Afghanistan                              8
Albania                                 23
Algeria                                 40
Andorra                                  0
Angola                                   2
                                        ..
Venezuela, Bolivarian Republic of...    28
Viet Nam                                78
Yemen                                    3
Zambia                                   4
Zimbabwe                                14
Name: LanguageWorkedWith, Length: 179, dtype: int64

In [60]:
country_grp["LanguageWorkedWith"].apply(lambda x: x.str.contains("Python").sum()).loc["India"]

3105

# Solving Problem

In [66]:
country_respondents = df["Country"].value_counts()
country_respondents

United States            20949
India                     9061
Germany                   5866
United Kingdom            5737
Canada                    3395
                         ...  
Brunei Darussalam            1
Papua New Guinea             1
Dominica                     1
Sao Tome and Principe        1
Saint Kitts and Nevis        1
Name: Country, Length: 179, dtype: int64

In [65]:
country_uses_python = country_grp["LanguageWorkedWith"].apply(lambda x: x.str.contains("Python").sum())
country_uses_python

Country
Afghanistan                              8
Albania                                 23
Algeria                                 40
Andorra                                  0
Angola                                   2
                                        ..
Venezuela, Bolivarian Republic of...    28
Viet Nam                                78
Yemen                                    3
Zambia                                   4
Zimbabwe                                14
Name: LanguageWorkedWith, Length: 179, dtype: int64

In [78]:
python_df = pd.concat([country_uses_python, country_respondents], axis="columns", sort=False)

In [82]:
python_df.sort_values(by = ["Country"],ascending=False)

LanguageWorkedWith  Country
United States                                  10083    20949
India                                           3105     9061
Germany                                         2451     5866
United Kingdom                                  2384     5737
Canada                                          1558     3395
...                                              ...      ...
North Korea                                        0        1
Niger                                              1        1
Saint Vincent and the Grenadines                   0        1
Chad                                               0        1
Brunei Darussalam                                  0        1

[179 rows x 2 columns]

In [87]:
python_df.rename(columns={"LanguageWorkedWith":"No knows Python", "Country":"No of Respondents"}, inplace=True)

In [90]:
python_df.sort_values(by = ["No knows Python", "No of Respondents"],ascending=False)

No knows Python  No of Respondents
United States                               10083              20949
India                                        3105               9061
Germany                                      2451               5866
United Kingdom                               2384               5737
Canada                                       1558               3395
...                                           ...                ...
North Korea                                     0                  1
Papua New Guinea                                0                  1
Saint Kitts and Nevis                           0                  1
Saint Vincent and the Grenadines                0                  1
Tonga                                           0                  1

[179 rows x 2 columns]

from Unites states, 10083 people responded that they knows python out of 20949

from India, 3105 people responded that they knows python out of 9061

In [91]:
python_df["perct_people"] = (python_df["No knows Python"]/ python_df["No of Respondents"]) * 100

In [95]:
python_df.sort_values(by = ["perct_people"],ascending=False)

No knows Python  No of Respondents  \
Dominica                                        1                  1   
Niger                                           1                  1   
Timor-Leste                                     1                  1   
Sao Tome and Principe                           1                  1   
Turkmenistan                                    6                  7   
...                                           ...                ...   
Tonga                                           0                  1   
North Korea                                     0                  1   
Papua New Guinea                                0                  1   
Tajikistan                                      0                  7   
Lao People's Democratic Republic                0                  3   

                                  perct_people  
Dominica                            100.000000  
Niger                               100.000000  
Timor-Leste                         100.000000  
Sao Tome and Principe               100.000000  
Turkmenistan                         85.714286  
...                                        ...  
Tonga                                 0.000000  
North Korea                           0.000000  
Papua New Guinea                      0.000000  
Tajikistan                            0.000000  
Lao People's Democratic Republic      0.000000  

[179 rows x 3 columns]

In [96]:
python_df.sort_values(by = ["No knows Python", "No of Respondents"], ascending=False, inplace=True)

In [97]:
python_df

No knows Python  No of Respondents  \
United States                               10083              20949   
India                                        3105               9061   
Germany                                      2451               5866   
United Kingdom                               2384               5737   
Canada                                       1558               3395   
...                                           ...                ...   
North Korea                                     0                  1   
Papua New Guinea                                0                  1   
Saint Kitts and Nevis                           0                  1   
Saint Vincent and the Grenadines                0                  1   
Tonga                                           0                  1   

                                  perct_people  
United States                        48.131176  
India                                34.267741  
Germany                              41.783157  
United Kingdom                       41.554820  
Canada                               45.891016  
...                                        ...  
North Korea                           0.000000  
Papua New Guinea                      0.000000  
Saint Kitts and Nevis                 0.000000  
Saint Vincent and the Grenadines      0.000000  
Tonga                                 0.000000  

[179 rows x 3 columns]

In [98]:
python_df.loc["India"]

No knows Python      3105.000000
No of Respondents    9061.000000
perct_people           34.267741
Name: India, dtype: float64